### DISCUSSION TITLE
test-execution-chart for Maven

### DISCUSSION THREAD

**Discussion by sobelek** [platan/tests-execution-chart#53] [Ideas]
Hey.
@platan 

I recently wanted to use test-execution-chart in Maven project and to my suprise it's gradle only plugin.
I am thinking about implementing the same functionality for maven and wanted to discuss possible solutions.

Do you think it would be feasible to do multi build system project or it would be better to either split it into totally separate project with shared common code?

**Comment by platan** [comment]
Hi @sobelek. Great idea. Think both approaches can be used. No matter which one we choose this project must be refactored to have at least two modules - first with common code (independent from Gradle API) and second with plugin (Gradle API). If you want I can prepare these modules. Let me know if you want. 

One repo with two plugins allows to keep both plugins synchronized with all features. 
On the other hand a separate repo for new plugin should be easier to set up. One option is to start with a separate repo and then merge both projects. 

I'm fine with both approaches. Feel free to choose the one which is better for you.

**Reply by sobelek** [reply]
Awesome! Let's start with separate repo. You can start working on splitting current code into 2 modules and I will start reaserching maven plugin API. I already started it a little bit and so far it looks like there is no way of getting `start` and `endtime` of test. Only `duration`. So possibly we would also need to rework plugin API to allows:
- saving test `start` time separetly and storing it
- later on saving `duration`
- at the end merging it

**Comment by platan** [comment]
> it looks like there is no way of getting `start` and `endtime` of test. Only `duration`. So possibly we would also need to rework plugin API to allows:
> * saving test `start` time separetly and storing it
> * later on saving `duration`

Just to be clear. There is no way to get `start` and `end` of test at once but there is a way to get `start` and `duration` in a separate steps? `start`+`end` or `start`+`duration` are crucial to draw a chart. Just `duration` is not enough.

**Comment by platan** [comment]
I've started work on extracting common code. You can see current results here https://github.com/platan/tests-execution-chart/compare/split-api?expand=1

commons module is published to https://github.com/features/packages. You can see it here
https://github.com/platan/tests-execution-chart/packages/1933174. 

I'm using it for the first time and I think it's not the best solution, since you have to authenticate to GH if you want to use artifacts from it ([docs](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-gradle-registry#using-a-published-package)). 
`build.gradle.kts`:
```kt
repositories {
    maven {
        url = uri("https://maven.pkg.github.com/platan/tests-execution-chart")
        credentials {
            username = System.getenv("USERNAME")
            password = System.getenv("TOKEN")
        }
    }
}

dependencies {
    implementation("io.github.platan:tests-execution-chart-commons:0.1.0-SNAPSHOT")
}
```
But it works. In the future I plan to publish it to Maven Central Repository.

**Comment by platan** [comment]
[tests-execution-chart-commons](https://github.com/platan/tests-execution-chart/tree/main/tests-execution-chart-commons) module is in master. Packages are currently available via https://github.com/platan/tests-execution-chart/packages/.

**Comment by platan** [comment]
Snapshots of are available in https://s01.oss.sonatype.org/content/repositories/snapshots.
